# Testing RNAseq analysis
https://gist.github.com/jdblischak/11384914
# Install packages:----
source("https://bioconductor.org/biocLite.R")
biocLite("edgeR", dependencies = TRUE)
# a (update all)
# y (write personal library)

install.packages("ggplot2")

library(edgeR)
library(ggplot2)
##
# Download data:----
fname <- "http://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE49712&format=file&file=GSE49712_HTSeq.txt.gz"
download.file(fname, destfile = "GSE49712_HTSeq.txt.gz")

##
# Load data:----
data_raw <- read.table("GSE49712_HTSeq.txt.gz", header = TRUE)

##
# Quality control:----
class(data_raw)
dim(data_raw)
# 21716 genes, 10 samples
head(data_raw)
tail(data_raw)


# Data includes two different groups for comparison.
# Group A is five technical replicates of an equal mixture of RNA from ten different human cell lines.
# Group B is five technical replicates of pooled RNA from several regions of the brain.
# [a]: http://www.chem.agilent.com/Library/usermanuals/Public/740000.pdf
# [b]: http://www.lifetechnologies.com/order/catalog/product/AM6050

# Unlike the WGS data we saw before which generated aligned reads and variant calls,
# RNAseq generates count data where the number of reads aligned to a gene is counted and interpreted as expression level.
# Differential expression analysis allows inference of whether genes have higher 
# or lower expression under different conditions, between sample comparisons, across time.
# There is usually some comparison in mind, but not always. 
# RNAseq can also be used to describe the profile of a tissue, cell, or organism.

# The last 5 lines have summary information that needs to be removed prior to testing.
data_clean <- data_raw[1:(nrow(data_raw) - 5),]
dim(data_clean)

# Remove unexpressed and low-expressed genes:
# One method is to choose a cutoff that is informed by the data
# Calculate the median log-2 counts per gene per millian mapped reads (cpm)

cpm_log <- cpm(data_clean, log = TRUE)
median_log2_cpm <- apply(cpm_log, 1, median)
hist(median_log2_cpm)
expr_cutoff <- -1
abline(v = expr_cutoff, col = "red", lwd = 3)
sum(median_log2_cpm > expr_cutoff)
# 16217 genes retained
data_clean <- data_clean[median_log2_cpm > expr_cutoff, ]

# Check that we've filtered our genes:
dim(data_clean)
# 16217 genes, 10 samples
cpm_log_clean <- cpm(data_clean, log = TRUE)
median_log2_cpm_clean <- apply(cpm_log_clean, 1, median)
hist(median_log2_cpm_clean)

# Generate a heatmap of the correlation matrix to examine relationships between the comparison groups
# In this case, Group A and Group B are very different groups of cell lines
# so we expect them to have different expression profiles--overall gene expression patterns
# In addition, because the 5 sample replicates are technical replicates (sequencing the same sample 5 times)
# the within-group variation in gene expression is expected to be very low
heatmap(cor(cpm_log_clean))

# It is also common practice to use principal components analysis to look at relationships between comparison groups.
pca <- prcomp(t(cpm_log_clean), scale. = TRUE ) # transposes the data frame, scales variables so that they have unit variance
PCA_data <- pca$x
head(PCA_data)

plot(pca$x[, 1], pca$x[, 2], pch = ".", xlab = "PC1", ylab = "PC2")
text(pca$x[, 1], pca$x[, 2], labels = colnames(cpm_log))
summary(pca)

###

# Two group comparison:----
# Preparing the data for DE, we need to assign a grouping variable that identifies
# the parts of the count data that belong to Group A and B
# The group variable and the count data are used to generate a DGEList
# which is used by edgeR to store data from a DE experiment

head(data_clean) # We want only the first letter of the column name
group <- substr(colnames(data_clean), 1, 1) # start at 1st position, end with 1st position
group

y <- DGEList(counts = data_clean, group = group)
y
class(y)

# Normalize gene counts:
# Used the TMM method (trimmed means of m values).
# Read counts for moderately to lowly expressed genes can be strongly influenced by small fluctuations in the expression of highly expressed genes.
# Small differences in highly expressed genes can make it look like there are also lots of lowly expressed genes with differential expression.
# TMM removes very high and very low expressed genes as well as genes that are very different across samples.
# The total counts for the subset of genes are then scaled.
# Assumes that the majority of genes are not different between any two samples

y <- calcNormFactors(y)
y$samples


# Model variance of read counts per gene:
# Measurements of gene expression variance generated by RNAseq data are typically overdispersed (larger than the mean gene expression).
# For this reason a negative binomial distribution is used.
# edgeR treats the Poisson variance as simple sampling variance with the dispersion estimate as the biological coefficient of variation.
# Technical biases are also included in this estimate.
# edgeR then determines a common dispersion estimate.
# Then, models the mean-variance relationship using the dispersion information.
# Calculates a dispersion estimate per gene and shrinks it towards the trended dispersion.
# Gene-specific (tagwise) dispersion estimates are used in the DE test

y <- estimateDisp(y)








